Assigning null to a const declaration annotated with `anyref` crashes the compiler

<!--
 Thanks for submitting an issue to AssemblyScript! Please take a moment to
 read the contributing guidelines linked below to get off to a good start 🙂
-->

## Code

```ts
// the `const` here is necessary
const a: anyref = null;
```

[Try it in the playground](https://www.assemblyscript.org/editor.html#IyFvcHRpbWl6ZT1zcGVlZCZydW50aW1lPXN0dWIKY29uc3QgYTogYW55cmVmID0gbnVsbDsKIyFodG1sCjx0ZXh0YXJlYSBpZD0ib3V0cHV0IiBzdHlsZT0iaGVpZ2h0OiAxMDAlOyB3aWR0aDogMTAwJSIgcmVhZG9ubHk+PC90ZXh0YXJlYT4KPHNjcmlwdD4KbG9hZGVyLmluc3RhbnRpYXRlKG1vZHVsZV93YXNtLCB7IC8qIGltcG9ydHMgKi8gfSkKICAudGhlbigoeyBleHBvcnRzIH0pID0+IHsKICAgIGNvbnN0IG91dHB1dCA9IGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCdvdXRwdXQnKQogICAgZm9yIChsZXQgaSA9IDA7IGkgPD0gMTA7ICsraSkgewogICAgICBvdXRwdXQudmFsdWUgKz0gYGZpYigke2l9KSA9ICR7ZXhwb3J0cy5maWIoaSl9XG5gCiAgICB9CiAgfSkKPC9zY3JpcHQ+)

## Stack trace

```
▌ Whoops, the AssemblyScript compiler has crashed during compile :-(
▌ 
▌ Here is the stack trace hinting at the problem, perhaps it's useful?
▌ 
▌ AssertionError: assertion failed
▌     at assert (D:\assemblyscript\std\portable\index.js:200:11)
▌     at Module.runExpression (D:\assemblyscript\src\module.ts:2574:7)
▌     at Compiler.compileGlobal (D:\assemblyscript\src\compiler.ts:1272:34)
▌     at Compiler.compileTopLevelStatement (D:\assemblyscript\src\compiler.ts:2140:20)
▌     at Compiler.compileFile (D:\assemblyscript\src\compiler.ts:1111:12)
▌     at Compiler.compile (D:\assemblyscript\src\compiler.ts:486:14)
▌     at Object.compile (D:\assemblyscript\src\index.ts:280:32)
▌     at D:\assemblyscript\cli\asc.js:902:31
▌     at measure (D:\assemblyscript\cli\asc.js:1409:3)
▌     at Object.main (D:\assemblyscript\cli\asc.js:900:24)
▌     at runTest (D:\assemblyscript\tests\compiler.js:178:7)
▌     at Array.forEach (<anonymous>)
▌     at D:\assemblyscript\tests\compiler.js:557:16
▌     at processTicksAndRejections (node:internal/process/task_queues:96:5)
▌ 
▌ If it refers to the dist files, try to 'npm install source-map-support' and
▌ run again, which should then show the actual code location in the sources.
▌ 
▌ If you see where the error is, feel free to send us a pull request. If not,
▌ please let us know: https://github.com/AssemblyScript/assemblyscript/issues
▌ 
▌ Thank you!
```

--- a/src/builtins.ts
+++ b/src/builtins.ts
@@ -907,7 +907,7 @@ function builtin_isConstant(ctx: BuiltinContext): ExpressionRef {
   var expr = compiler.compileExpression(ctx.operands[0], Type.auto);
   compiler.currentType = Type.bool;
   if (!mustPreserveSideEffects(expr, module.ref)) {
-    return module.i32(getExpressionId(expr) == ExpressionId.Const ? 1 : 0);
+    return module.i32(module.isConstExpression(expr) ? 1 : 0);
   }
   return module.block(null, [
     module.maybeDrop(expr),

--- a/src/compiler.ts
+++ b/src/compiler.ts
@@ -1158,16 +1158,14 @@ export class Compiler extends DiagnosticEmitter {
         this.currentFlow = previousFlow;
       }
 
-      // If not a constant, attempt to precompute
-      if (getExpressionId(initExpr) != ExpressionId.Const) {
+      // If not a constant expression, attempt to precompute
+      if (!module.isConstExpression(initExpr)) {
         if (isDeclaredConstant) {
-          if (getExpressionId(initExpr) != ExpressionId.Const) {
-            let precomp = module.runExpression(initExpr, ExpressionRunnerFlags.PreserveSideeffects);
-            if (precomp) {
-              initExpr = precomp;
-            } else {
-              initializeInStart = true;
-            }
+          let precomp = module.runExpression(initExpr, ExpressionRunnerFlags.PreserveSideeffects);
+          if (precomp) {
+            initExpr = precomp;
+          } else {
+            initializeInStart = true;
           }
         } else {
           initializeInStart = true;

--- a/src/module.ts
+++ b/src/module.ts
@@ -2582,7 +2582,7 @@ export class Module {
     var runner = binaryen._ExpressionRunnerCreate(this.ref, flags, maxDepth, maxLoopIterations);
     var precomp =  binaryen._ExpressionRunnerRunAndDispose(runner, expr);
     if (precomp) {
-      if (!this.isConstExpression(precomp)) return expr;
+      if (!this.isConstExpression(precomp)) return 0;
       assert(getExpressionType(precomp) == getExpressionType(expr));
     }
     return precomp;

--- a/src/module.ts
+++ b/src/module.ts
@@ -2582,10 +2582,33 @@ export class Module {
     var runner = binaryen._ExpressionRunnerCreate(this.ref, flags, maxDepth, maxLoopIterations);
     var precomp =  binaryen._ExpressionRunnerRunAndDispose(runner, expr);
     if (precomp) {
-      assert(getExpressionId(precomp) == ExpressionId.Const);
+      if (!this.isConstExpression(precomp)) return expr;
       assert(getExpressionType(precomp) == getExpressionType(expr));
     }
     return precomp;
+  }
+
+  isConstExpression(expr: ExpressionRef, features: FeatureFlags = 0): bool {
+    switch (getExpressionId(expr)) {
+      case ExpressionId.Const:
+      case ExpressionId.RefNull:
+      case ExpressionId.RefFunc:
+      case ExpressionId.I31New: return true;
+      case ExpressionId.Binary: {
+        if (this.getFeatures() & FeatureFlags.ExtendedConst) {
+          switch (getBinaryOp(expr)) {
+            case BinaryOp.AddI32:
+            case BinaryOp.SubI32:
+            case BinaryOp.MulI32:
+            case BinaryOp.AddI64:
+            case BinaryOp.SubI64:
+            case BinaryOp.MulI64: return this.isConstExpression(getBinaryLeft(expr)) && this.isConstExpression(getBinaryRight(expr));
+          }
+        }
+        break;
+      }
+    }
+    return false;
   }
 
   // source map generation

--- a/tests/compiler/features/gc.debug.wat
+++ b/tests/compiler/features/gc.debug.wat
@@ -2,6 +2,7 @@
  (type $none_=>_none (func))
  (type $i32_i32_i32_i32_=>_none (func (param i32 i32 i32 i32)))
  (import "env" "abort" (func $~lib/builtins/abort (param i32 i32 i32 i32)))
+ (global $features/gc/a anyref (ref.null any))
  (global $~lib/memory/__data_end i32 (i32.const 60))
  (global $~lib/memory/__stack_pointer (mut i32) (i32.const 16444))
  (global $~lib/memory/__heap_base i32 (i32.const 16444))
@@ -10,6 +11,7 @@
  (data (i32.const 12) ",\00\00\00\00\00\00\00\00\00\00\00\01\00\00\00\1c\00\00\00f\00e\00a\00t\00u\00r\00e\00s\00/\00g\00c\00.\00t\00s\00")
  (table $0 1 funcref)
  (elem $0 (i32.const 1))
+ (export "a" (global $features/gc/a))
  (export "memory" (memory $0))
  (export "_start" (func $~start))
  (func $features/gc/test_i31

--- a/tests/compiler/features/gc.release.wat
+++ b/tests/compiler/features/gc.release.wat
@@ -1,8 +1,10 @@
 (module
  (type $none_=>_none (func))
+ (global $features/gc/a anyref (ref.null any))
  (memory $0 1)
  (data (i32.const 1036) ",")
  (data (i32.const 1048) "\01\00\00\00\1c\00\00\00f\00e\00a\00t\00u\00r\00e\00s\00/\00g\00c\00.\00t\00s")
+ (export "a" (global $features/gc/a))
  (export "memory" (memory $0))
  (export "_start" (func $~start))
  (func $~start

--- a/tests/compiler/features/gc.ts
+++ b/tests/compiler/features/gc.ts
@@ -6,3 +6,9 @@ function test_i31(): void {
 }
 
 test_i31();
+
+// constant globals
+
+export const a: anyref = null;
+// export const b: i31ref = null; // TODO: not yet nullable in Binaryen
+// export const c: dataref = null; // TODO: not yet nullable in Binaryen

--- a/tests/compiler/features/reference-types.debug.wat
+++ b/tests/compiler/features/reference-types.debug.wat
@@ -13,7 +13,9 @@
  (global $features/reference-types/externGlobalInit (mut externref) (ref.null extern))
  (global $features/reference-types/anyGlobal (mut anyref) (ref.null any))
  (global $features/reference-types/anyGlobalInit (mut anyref) (ref.null any))
- (global $features/reference-types/otherFuncGlobal (mut funcref) (ref.null func))
+ (global $features/reference-types/otherFuncGlobal (mut funcref) (ref.func $features/reference-types/someFunc))
+ (global $features/reference-types/a externref (ref.null extern))
+ (global $features/reference-types/b funcref (ref.null func))
  (global $~lib/memory/__data_end i32 (i32.const 92))
  (global $~lib/memory/__stack_pointer (mut i32) (i32.const 16476))
  (global $~lib/memory/__heap_base i32 (i32.const 16476))
@@ -26,6 +28,8 @@
  (export "somethingReal" (func $features/reference-types/somethingReal))
  (export "somethingNull" (func $features/reference-types/somethingNull))
  (export "internal" (func $features/reference-types/internal))
+ (export "a" (global $features/reference-types/a))
+ (export "b" (global $features/reference-types/b))
  (export "memory" (memory $0))
  (start $~start)
  (func $features/reference-types/testLocal<funcref>
@@ -268,8 +272,6 @@
    call $~lib/builtins/abort
    unreachable
   end
-  ref.null func
-  global.set $features/reference-types/funcGlobalInit
   global.get $features/reference-types/funcGlobalInit
   ref.is_null
   i32.eqz
@@ -311,8 +313,6 @@
    call $~lib/builtins/abort
    unreachable
   end
-  ref.null extern
-  global.set $features/reference-types/externGlobalInit
   global.get $features/reference-types/externGlobalInit
   ref.is_null
   i32.eqz
@@ -354,8 +354,6 @@
    call $~lib/builtins/abort
    unreachable
   end
-  ref.null any
-  global.set $features/reference-types/anyGlobalInit
   global.get $features/reference-types/anyGlobalInit
   ref.is_null
   i32.eqz
@@ -384,8 +382,6 @@
    call $~lib/builtins/abort
    unreachable
   end
-  ref.func $features/reference-types/someFunc
-  global.set $features/reference-types/otherFuncGlobal
   global.get $features/reference-types/otherFuncGlobal
   ref.is_null
   if

--- a/tests/compiler/features/reference-types.release.wat
+++ b/tests/compiler/features/reference-types.release.wat
@@ -9,6 +9,8 @@
  (import "reference-types" "external" (func $features/reference-types/external (param externref) (result externref)))
  (global $features/reference-types/funcGlobal (mut funcref) (ref.null func))
  (global $features/reference-types/anyGlobal (mut anyref) (ref.null any))
+ (global $features/reference-types/a externref (ref.null extern))
+ (global $features/reference-types/b funcref (ref.null func))
  (memory $0 1)
  (data (i32.const 1036) "L")
  (data (i32.const 1048) "\01\00\00\006\00\00\00f\00e\00a\00t\00u\00r\00e\00s\00/\00r\00e\00f\00e\00r\00e\00n\00c\00e\00-\00t\00y\00p\00e\00s\00.\00t\00s")
@@ -17,6 +19,8 @@
  (export "somethingReal" (func $features/reference-types/somethingReal))
  (export "somethingNull" (func $features/reference-types/somethingNull))
  (export "internal" (func $features/reference-types/internal))
+ (export "a" (global $features/reference-types/a))
+ (export "b" (global $features/reference-types/b))
  (export "memory" (memory $0))
  (start $~start)
  (func $features/reference-types/someFunc

--- a/tests/compiler/features/reference-types.ts
+++ b/tests/compiler/features/reference-types.ts
@@ -99,3 +99,8 @@ assert(otherFuncGlobal);
 // can assign any reference type to anyref
 
 anyGlobal = funcGlobal;
+
+// constant globals
+
+export const a: externref = null;
+export const b: funcref = null;

